src/sentry/event_manager.py

```diff
@@ -559,8 +560,8 @@ def save(self, project_id, raw=False, assume_normalized=False):
         # We need to swap out the data with the one internal to the newly
         # created event object
         event = self._get_event_instance(project_id=project_id)
-        self._data = data = event.data.data
-
+        # self._data = data = event.data.data
+        assert self._data == data == event.data.data
```

Comment:
Overall I am not sure I get why we need to change the type of Event.data from EventDict to NodeData.
Also what's the reason for these three representations of event to coexist ? Are we modifying them independently in such a way that we still have to preserve all of them ?

Reply:
My goal is for accessing event.data to stay unchanged. Prior to this change NodeField took care of always returning you a NodeData instance https://github.com/getsentry/sentry/blob/master/src/sentry/db/models/fields/node.py#L219.

I'm not sure if we'll be able to do away with any of this in future, likely not anytime soon though since RawEvent still uses NodeData in the "old" way

Reply:
The type of event.data was always `NodeData`. `EventDict` was wrapped by NodeData.

The only purpose of `EventDict` was to apply renormalization transparently when `NodeField` was loading the data from nodestore. I would love to get rid of this implicit hack if we know all the places where `event.data` is accessed. The assumption at the time was that we don't because of plugins.

Please be very careful when rolling this out. Applying renormalization too often can cause a slowdown of post-process-forwarder and create a backlog that impacts how fast events are stored and indexed. We have observed such a backlog, and concluded that pickling and unpickling (or something like it) of the event object causes renormalization to be applied again when it was not necessary. That's why sometimes we explicitly re-set event.data to itself with the `skip_renormalization` flag.

The other thing that can happen during deploy is that we do not renormalize when we should. That can cause UI crashes when old, partially broken events (that violate the schema, or an invariant in the data that the UI relies on). This is not a big problem because there are not many events like that, but it's hard to monitor for. More often the UI does not crash, but the API endpoint for the issue details page. That one is usually easy to monitor and easy to fix.

ccing @mitsuhiko on this because he has a better understanding of how NodeData works.